### PR TITLE
feat: per-repo config overrides (workflows, agents, defaults, model roles)

### LINKED ISSUE #29: feat: per-repo config overrides (workflows, agents, defaults, model roles)
Part of #23

## Context

**User Story:** As a user managing multiple repos with different development patterns, I want per-repo configuration overrides so each repo can have its own workflows, agents, and defaults without affecting other repos.

Currently `workflow_selection`, `defaults`, `agents_dir`, and `workflows_dir` are global in `~/.adl/config.yaml`. All repos share the same workflow routing, iteration limits, and agent definitions. This doesn't work when, for example, one repo needs a `security_audit` default workflow while another needs `feature`, or one repo needs stricter review cycles than another.

## Complexity Hint

**Subsystems touched:** config model (`models.py`), config loader (`config.py`), daemon (`main.py`), workflow engine consumers
**Independent stories:** 1
**Structural changes needed:** Yes — `RepoConfig` gains optional override fields, config resolution needs a merge layer

## Proposed Solution

Extend `RepoConfig` with optional override fields that merge with global config at runtime.

### Config shape

```yaml
# Global defaults
defaults:
  agents_dir: ./agents
  workflows_dir: ./workflows
  max_dev_cycles: 5
  max_review_cycles: 5
  external_reviewers: [gemini]

workflow_selection:
  default: feature
  label_map:
    bug: bug_fix

model_roles:
  smol: claude-haiku-4-5
  default: claude-sonnet-4-5
  slow: claude-opus-4-5

repos:
  - path: ~/projects/web-app
    project_number: 1
    # Uses all global defaults — no overrides

  - path: ~/projects/security-tool
    project_number: 2
    # Per-repo overrides
    agents_dir: ./custom-agents
    workflows_dir: ./custom-workflows
    defaults:
      max_dev_cycles: 3
      max_review_cycles: 10
      external_reviewers: [gemini, codex]
    workflow_selection:
      default: security_audit
      label_map:
        bug: bug_fix
        feature: security_feature
    model_roles:
      slow: claude-opus-4-5    # use opus for all review in this repo
```

### Resolution order

Per-repo fields override global fields via shallow merge:
1. Start with global `defaults` / `workflow_selection` / `model_roles`
2. If repo has the same field, repo value wins
3. For nested dicts (`label_map`, `model_roles`), merge keys — repo keys override global keys with the same name

### Runtime API

Add a `resolved_config(repo_cfg: RepoConfig, global_cfg: Config) -> ResolvedRepoConfig` function that returns a flat config with all overrides applied. Callers get a single object — no need to check two places.

## Acceptance Criteria

- [ ] VERIFY: `RepoConfig` accepts optional `defaults`, `workflow_selection`, `model_roles`, `agents_dir`, `workflows_dir` fields
- [ ] VERIFY: `config.py` parses per-repo override fields from YAML
- [ ] VERIFY: Omitted per-repo fields fall back to global config values
- [ ] VERIFY: Per-repo `defaults` dict is shallow-merged with global `defaults` (repo keys win)
- [ ] VERIFY: Per-repo `workflow_selection.label_map` is merged with global `label_map` (repo keys win)
- [ ] VERIFY: Per-repo `model_roles` is merged with global `model_roles` (repo keys win)
- [ ] VERIFY: A `resolved_config()` function returns a single merged config for a given repo
- [ ] VERIFY: Daemon passes resolved config (not global) to `process_issue` and downstream consumers
- [ ] VERIFY: Repos without overrides behave identically to current behavior
- [ ] VERIFY: `adl validate` reports per-repo override errors clearly (e.g., invalid workflow name)
- [ ] VERIFY: All existing tests pass

### Files to Modify

- `src/auto_dev_loop/models.py` — add optional override fields to `RepoConfig`; create `ResolvedRepoConfig` dataclass
- `src/auto_dev_loop/config.py` — parse per-repo override fields; add `resolved_config()` merge function
- `src/auto_dev_loop/main.py` — call `resolved_config()` per repo in `run_poll_cycle` and pass to `_process_issue_task`
- `src/auto_dev_loop/orchestrator.py` — ensure it uses the passed config (not a global), already should be the case
- `src/auto_dev_loop/cli.py` — `validate` command should validate per-repo overrides

### Approach

1. Add optional fields to `RepoConfig`:
   ```python
   @dataclass
   class RepoConfig:
       path: str
       project_number: int
       columns: dict[str, str] = ...
       owner: str | None = None
       # Overrides (None = use global)
       agents_dir: str | None = None
       workflows_dir: str | None = None
       defaults: Defaults | None = None
       workflow_selection: WorkflowSelectionConfig | None = None
       model_roles: dict[str, str] | None = None
   ```
2. Create `ResolvedRepoConfig` that flattens global + repo config into one object
3. Add `resolve_repo_config(repo: RepoConfig, global_cfg: Config) -> ResolvedRepoConfig` in `config.py`
4. Merge logic: `{**global_dict, **repo_dict}` for flat dicts; field-level override for dataclasses
5. Update `run_poll_cycle` to resolve config per repo before passing to issue processing
6. Update `validate` command to resolve and validate each repo's merged config

### Patterns to Follow

- Config loading in `config.py` already handles optional fields with `.get()` — extend the pattern
- `Defaults` dataclass uses `field(default_factory=...)` for mutable defaults — same pattern for repo overrides
- Keep `Config` as the global container; `ResolvedRepoConfig` is a per-repo view

### Constraints

- Must be fully backwards compatible — existing configs without per-repo overrides must work unchanged
- Merge must be deterministic — no ambiguity about which value wins
- `resolved_config()` must be pure (no side effects) — easy to test

### In Scope
- `RepoConfig` override fields
- Config parsing for per-repo sections
- `resolved_config()` merge function
- Daemon integration (pass resolved config downstream)
- Validation of merged configs

### Out of Scope
- Per-repo config files on disk (everything stays in `~/.adl/config.yaml`)
- `adl add` integration (will reference this in #25 once available)
- Config inheritance chains (only two levels: global → repo)
- GUI or web-based config editor

## Dependencies

- None — can be implemented independently
- #25 (`adl add`) should be updated to offer per-repo overrides during onboarding once this lands

## Testing Requirements

- [ ] Unit tests for `resolve_repo_config()` — global only, repo overrides, partial overrides
- [ ] Unit tests for dict merge behavior (repo keys override, global keys preserved)
- [ ] Unit tests for `load_config()` with per-repo override fields
- [ ] Test: config without overrides produces identical behavior to current
- [ ] Test: `validate` catches invalid per-repo workflow names

### PR DESCRIPTION
**Pull request by funny-otter**
## Summary

- Add optional override fields to `RepoConfig` (`agents_dir`, `workflows_dir`, `defaults`, `workflow_selection`, `model_roles`) that merge with global config at runtime
- Add `ResolvedRepoConfig` dataclass and pure `resolve_repo_config()` merge function
- Daemon resolves config per-repo before passing to `process_issue` — all downstream consumers work unchanged via `AppConfig` type alias
- `adl validate` reports per-repo override errors
- Full backwards compatibility — repos without overrides behave identically

Closes #29

## Test plan

- [x] 20 new tests covering model fields, YAML parsing, merge logic, purity, edge cases, CLI validation
- [x] Full suite passes (419 tests)
- [ ] Manual: add per-repo overrides to a config and run `adl validate`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Repository-specific configuration overrides (per-repo agent/workflow paths, defaults, workflow selection, model roles) and resolution/merge into effective repo configs.

* **Bug Fixes / Validation**
  * Enhanced validation with clearer errors and warnings for invalid per-repo settings and missing workflow/agent directories.

* **Compatibility**
  * Unified config type accepted across command surfaces to simplify downstream usage.

* **Tests**
  * Extensive tests covering parsing, merge behavior, validation, CLI checks, and resolution purity.

### CODE REVIEW

File: src/auto_dev_loop/main.py [hunk 1]
```diff
@@ -197,6 +197,7 @@ async def run_poll_cycle(
         completed = await store.list_terminal_issue_keys()
         state.completed_keys.update(completed)
 
+        resolved = resolve_repo_config(repo_cfg, config)
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/auto_dev_loop/main.py]: Fixed: resolve_repo_config() now wrapped in try/except per repo — one bad config no longer aborts the cycle

File: src/auto_dev_loop/cli.py [hunk 2]
```diff
@@ -53,10 +53,19 @@ def validate(
...
         cfg = load_config(config)
+        # Validate per-repo overrides by resolving each repo
+        for i, repo_cfg in enumerate(cfg.repos):
+            if not isinstance(repo_cfg, RepoConfig):
+                continue
+            try:
+                resolve_repo_config(repo_cfg, cfg)
+            except (TypeError, KeyError, ValueError, AttributeError, ConfigError) as e:
+                raise ConfigError(f"Invalid per-repo override in repos[{i}] ({repo_cfg.path}): {e}")
         typer.echo(f"Config OK: {len(cfg.repos)} repo(s), version {cfg.version}")
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/auto_dev_loop/cli.py]: Fixed: adl validate now loads workflows from disk and checks that workflow_selection.default, label_map values, and priority_overrides values reference existing workflow IDs — warns when workflows_dir is not accessible

**Comment by funny-otter** [comment]
## PR Comment Status

| Status | Threads | Review Bodies | Total |
|--------|---------|---------------|-------|
| Resolved | 0 | 2 | 2 |
| Fixed by DLC | 7 | 0 | 7 |
| Skipped (user decision) | 1 | 0 | 1 |
| Discussion (needs human) | 1 | 0 | 1 |
| Blocked | 0 | 0 | 0 |
| Dismissed | 0 | 0 | 0 |
| **Total** | **9** | **2** | **11** |

## Decisions

- `src/auto_dev_loop/main.py:200` — Fixed: wrapped `resolve_repo_config()` in try/except (4 threads from 3 reviewers)
- `src/auto_dev_loop/config.py:164` — Fixed: added isinstance checks for nested override fields
- `src/auto_dev_loop/config.py:196` — Fixed: copy mutable list fields to prevent shared references
- `src/auto_dev_loop/cli.py:69` — Fixed: `adl validate` now checks workflow_selection references against available workflows
- `src/auto_dev_loop/config.py:163` — Skipped: shallow merge kept for consistency with other dict merges
- `src/auto_dev_loop/models.py:197` — Discussion: Protocol vs union deferred — union works, both types in same file

## Follow-up

Author will address some remaining items manually.

Some remaining items deferred — out of scope for this PR.

File: src/auto_dev_loop/cli.py [hunk 3]
```diff
@@ -53,10 +53,77 @@ def validate(
...
+            except (TypeError, KeyError, ValueError, AttributeError, ConfigError) as e:
+                raise ConfigError(f"Invalid per-repo override in repos[{i}] ({repo_cfg.path}): {e}")
+
+            # Validate workflow_selection references against available workflows
+            wf_dir = Path(resolved.defaults.workflows_dir)
+            if not wf_dir.is_absolute():
+                repo_path = Path(repo_cfg.path)
+                if repo_path.is_dir():
+                    wf_dir = repo_path / wf_dir
+            if wf_dir.is_dir():
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/auto_dev_loop/cli.py]: Fixed: removed repo_path join — resolve_repo_config() already rebases overridden paths, eliminating the CWD resolution concern

File: src/auto_dev_loop/config.py [hunk 4]
```diff
@@ -112,3 +128,115 @@ def load_config(path: Path) -> Config:
...
+    if repo.defaults is not None:
+        if not isinstance(repo.defaults, dict):
+            raise ConfigError(
+                f"Per-repo defaults must be a mapping, "
+                f"got {type(repo.defaults).__name__} in {repo.path}"
+            )
+        valid_keys = set(Defaults.__dataclass_fields__)
+        for k, v in repo.defaults.items():
+            if k in valid_keys:
+                base[k] = v
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/auto_dev_loop/config.py]: Fixed: repo-provided list values are now copied via list(v) before assigning into base — preserves pure function contract

File: src/auto_dev_loop/config.py [hunk 5]
```diff
@@ -112,3 +128,121 @@ def load_config(path: Path) -> Config:
...
+        base["workflows_dir"] = repo.workflows_dir
+
+    # Rebase relative path overrides against repo.path so they resolve
+    # correctly regardless of daemon CWD.
+    _PATH_FIELDS = {"agents_dir", "workflows_dir"}
+    for pf in _PATH_FIELDS:
+        if pf in base and base[pf] != getattr(gd, pf):
+            raw_path = base[pf]
+            if not isinstance(raw_path, str):
+                raise ConfigError(
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/auto_dev_loop/config.py]: Fixed: path rebasing now tracks explicitly overridden fields via a set instead of value equality — same-value repo overrides are correctly rebased against repo.path

File: src/auto_dev_loop/config.py [hunk 6]
```diff
@@ -112,3 +128,121 @@ def load_config(path: Path) -> Config:
...
+            if not isinstance(prio_val, dict):
+                raise ConfigError(
+                    f"Per-repo workflow_selection.priority_overrides[{prio_key!r}] "
+                    f"must be a mapping, got {type(prio_val).__name__} in {repo.path}"
+                )
+        merged_ws = WorkflowSelectionConfig(
+            default=rws.get("default", gws.default),
+            label_map={**gws.label_map, **repo_label_map},
+            priority_overrides={k: dict(v) for k, v in {**gws.priority_overrides, **repo_priority}.items()},
+        )
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/auto_dev_loop/config.py]: Fixed: priority_overrides are now deep-merged per priority key — repo labels override matching global labels without dropping unrelated ones under the same priority
